subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java

```diff
@@ -154,9 +164,34 @@ public S3Object getResource(URI uri) {
                 }
             }
         }
+
+        return results;
+    }
+
+    private List<String> resolveDirectoryResourceNames(ObjectListing objectListing) {
+        List<String> results = new ArrayList<String>();
+        List<String> commonPrefixes = objectListing.getCommonPrefixes();
+        if(null != commonPrefixes) {
+            for(String prefix : commonPrefixes) {
+                String dirName = extractDirectoryName(prefix);
+                if(null != dirName) {
+                    results.add(dirName);
+                }
+            }
+        }
+
         return results;
     }
 
+    private String extractDirectoryName(String key) {
+        Matcher matcher = DIRNAME_PATTERN.matcher(key);
+        if (matcher.find()) {
+            // Add a "/" to differentiate directories from files
```

Comment:
The comment does not make much sense now, does it? Please rephrase or remove it.